scripts/redis_scan.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python
+"""
+Safely scan Redis instance and report key stats.
+
+Can also set the TTL for keys to facilitate removing data.
+"""
+import signal
+import sys
+import time
+
+import click
+import redis
+
+loop = True
+
+
+def signal_handler(signum, frame):
+    """Signal handler."""
+
+    global loop
+    print('Caught ctrl-c, finishing up.')
+    loop = False
+
+
+def get_size(client, key, key_type):
+    """Get size of key."""
+
+    size = -1
+    if key_type == 'string':
+        size = client.strlen(key)
+    elif key_type == 'zset':
+        size = client.zcard(key)
+    elif key_type == 'set':
+        size = client.scard(key)
+    elif key_type == 'list':
+        size = client.llen(key)
+    elif key_type == 'hash':
+        size = client.hlen(key)
+
+    return size
+
+
+@click.command()
+@click.option('--file', 'file_name', default='redis-stats.log')
+@click.option('--match', default=None)
+@click.option(
+    '--ttl',
+    'set_ttl',
+    default=None,
+    type=click.INT,
+    help="Set TTL if one isn't already set (-1 will remove TTL)",
+)
+@click.option('--host', required=True)
+@click.option('--port', type=click.INT, default=6379)
+@click.option('--db', type=click.INT, default=0)
+@click.option('--delay', type=click.FLOAT, default=0.1)
+@click.option('--print', 'print_it', is_flag=True)
+def run(host, port, db, delay, file_name, print_it, match, set_ttl=None):
+    """Run scan."""
+
+    if set_ttl is not None and match is None:
+        print('You must specify match when setting TTLs!')
+        sys.exit(1)
+
+    client = redis.Redis(host=host, port=port, db=db)
+
+    if match:
+        print(f'Scanning redis keys with match: {match}\n')
+    else:
+        print('Scanning all redis keys\n')
+
+    # This is a string because we want the `while` below to run, and the string
+    # will be correctly interpreted by `client.scan` as a request for a new
+    # cursor.
+    cursor = '0'
+
+    signal.signal(signal.SIGINT, signal_handler)
+
+    with open(file_name, 'w') as log_file:
+        while cursor != 0 and loop:
```

Comment:
Use `client.scan_iter()` instead. No need to mess with cursors manually.

Reply:
I'd rather not go that way. We want to control pressure on Redis by not issuing too many requests too fast, so we avoid starving other requests, and the in-between of `SCAN` iterations seems like a natural place to do that. We also want to flush `log_file` to disk often so that, in case something happens, we will have more accurate data on what happened.

Of course that can be achieved by other means, like counting number of keys processed and adding a delay and flushing the file every 1,000 keys, for example. But then, if I'm going to add that logic, might as well just keep the cursor logic we already have. Another reason to not do that is that, if the `match` used to look for a key has a very sparse occurrence, we could end up firing a lot of requests in rapid succession to Redis before we process 1,000 keys, so it doesn't guarantee that we won't starve other requests.

Makes sense?